Construct the TestSuiteResult as we get the testEnd event from QUnit

brandonaaron opened the pull request:
Builds up the `TestSuiteResult` on demand from the QUnit `testEnd` event. This provides a better way of getting the `actual` and `expected` props into the @web/test-runner test results. This results in only allowing us to report a single failed assertion per a unit test though.

<img width="663" alt="Screenshot 2024-02-18 at 8 56 45 AM" src="https://github.com/brandonaaron/web-test-runner-qunit/assets/13718/e7eca040-293a-4343-ad18-210e48d6a50b">

brandonaaron committed:
Construct the TestSuiteResult as we get the testEnd event from QUnit

brandonaaron committed:
Use JSON.stringify as @rnixx suggested

rnixx commented on src/autorun.ts:
```diff
@@ -116,57 +123,89 @@ async function setupQUnit (qunitBasePath: string) {
-      errors: testResultErrors,
-      testResults
-    }).catch((err) => console.error(err))
+    // using a setTimeout here to try and help capture any out of sync assertions prior to calling sessionFinished
```
I don't really understand how a delay of 1 millisecond helps here. Do you have an example what case the timeout is trying to fix?

brandonaaron replied:
It was an attempt to capture [this type of failure](https://github.com/qunitjs/qunit/blob/13f0926119781c55eb4c4512b6aa58ecd7f7b7dd/src/test.js#L553). However, I think it can probably be dropped because we can't really know what a reasonable amount of time to wait is. I think not capturing these types of failures will just be a limitation of using @web/test-runner. Something to include in the readme.

```
QUnit.test('testing', assert => {
  setTimeout(() => {
    assert.true(false, 'test')
  }, 1)
})
```

rnixx replied:
i think you can safely ignore this case because here the test is broken. The correct test case would look something like:

    QUnit.test('testing', assert => {
        const done = assert.async();
        setTimeout(() => {
            assert.true(false, 'test');
            done();
        }, 1)
    })

brandonaaron committed:
Remove attempt to delay results for any lingering output

More specifically this was meant to look at capturing assertions that
happened after the test finished. However we do not know what a
reasonable amount of time is to delay the results of the unit tests.

brandonaaron committed:
Do not clobber the tests/assertions if a module/test is re-opened

Take for example the following:

```QUnit.module('Module A', () => {
  QUnit.test('test 1', assert => {
    assert.true(true, 'assertion 1')
  })
  QUnit.test('test 2', assert => {
    assert.true(true, 'assertion 2')
  })
})
QUnit.module('Module A', () => {
  QUnit.test('test 1', assert => {
    assert.true(true, 'assertion 3')
  })
})```

This will result in the following summary results:

```
 Module A [Chrome]
  test 1 [Chrome]
    ✓ assertion 1
    ✓ assertion 3
  test 2 [Chrome]
    ✓ assertion 2
```

brandonaaron committed:
include a basic async test

brandonaaron committed:
Add documentation explaining the construction of TestSuiteResult

brandonaaron committed:
Add gotchas/limitations section to README

rnixx commented on README.md:
````diff
@@ -28,3 +28,46 @@ export default {
+    ✓ assertion 3
+  test 2 [Chrome]
+    ✓ assertion 2
+```
````
If i skip the assertion description (e.g. assert.true(true)), the test name is printed. Is it really desired to print all assertions from tests in the summary? Or would it make more sense to just mark the test failed and skip the assertions in the summary? the error of the affected assertion can be included anyways. What do you think?

brandonaaron replied:
It has been a bit of a struggle to find the best mapping to the @web/test-runner data. If we go back to mapping the `QUnit.test` to the [`TestResult`](https://github.com/modernweb-dev/web/blob/f7fcf29cb79e82ad5622665d76da3f6b23d0ef43/packages/test-runner-core/src/test-session/TestSession.ts#L24) then we can only report a single failure at a time. It seemed better to map each assertion inside of a `QUnit.test` to a `TestResult` instead in order to capture all failed assertions within a test.

I'm honestly unsure which compromise is better: one failed assertion reporting per test per a run or reporting on every assertion as individual tests.

rnixx replied:
You mean a single assertion failure in a single qunit.test() block? It depends on the philosophy of testing. Very often subsequent tests inside a test block depend on prior assertions working. This "fail early" approach not clutters the report with subsequent faults and make it easier to find the initial error cause. At least i prefer this behavior.

brandonaaron replied:
Maybe It'd be nicer to output a default message if the assertion does not have one instead of the test name. Something like `assertion [index + 1]`.

rnixx replied:
I think we talk about 2 different things right now. The one thing is the printing of the error with a diff, which in my opinion should happen for the first failed assertion inside each qunit.test() block. This error also points me to the right code line via stack information. The other thing is reporting every assertion in the summary, with the "wrong" assertion name currently. I know qunit web UI also prints every assertion, but i think this is not necessary, the information which test block failed/succeeded should be enough information. But actually i am not religious on this topic.

brandonaaron replied:
Hmm... thinking about this some more. In my [original implementation](https://github.com/brandonaaron/web-test-runner-qunit/blob/2e81cba5b5ab335a6ca16c0513ce314893a0722b/src/autorun.ts#L121) I treated each assertion failure as a [global error](https://github.com/modernweb-dev/web/blob/f7fcf29cb79e82ad5622665d76da3f6b23d0ef43/packages/test-runner-core/src/test-session/TestSession.ts#L32). With that approach the `QUnit.test` still appropriately maps to a `TestResult` and we'll just skip associating the `TestResultError` and instead push it to the `errors` on the `TestSession`.

Yeah, I think there is a happy blend between this branch and the main branch that gets us where we want to be. I'll try to make some time later today to keep iterating on this.

brandonaaron replied:
I worked on this a bit just now but unfortunately @web/test-runner does not output the expected/actual diff for the [`TestSession.errors`](https://github.com/modernweb-dev/web/blob/f7fcf29cb79e82ad5622665d76da3f6b23d0ef43/packages/test-runner-core/src/test-session/TestSession.ts#L32).

This is a screen grab of a session where it is showing how it outputs the errors. I also used `console.log` to output what I'm giving @web/test-runner.

<img width="975" alt="Screenshot 2024-02-14 at 8 09 10 AM" src="https://github.com/brandonaaron/web-test-runner-qunit/assets/13718/aa56e7c7-84c3-4195-a4c3-8500f4e40387">

This puts us back to compromising on displaying all assertions or only exposing the first assertion failure in a test. I was initially leaning more towards mapping all QUnit assertions to an @web/test-runner `TestResult`. However, it is a bit simpler to just map the QUnit.test to the `TestResult` and only show the first failing assertion. In the future there could be a config item that would change the behavior to instead show all assertions as initially done in this PR.

brandonaaron replied:
I went ahead and implemented the change to map each `QUnit.test` to a `TestResult` and only show the first failed assertion.

rnixx replied:
Maybe is miss something completely, but why are you not adding the assertion failures to testResults? The missing diff output when adding them to errors was the reason for my initial issue report ;)

brandonaaron replied:
I might not be conveying the information very well. While iterating through ideas/solutions I was mentioning that @web/test-runner does not show the actual/expected diff when sending errors as TestSession.errors. However, in the latest commit this does pass along the assertion failure (but only one per test).

Given the following tests:

```
QUnit.test('testing 1', assert => {
  assert.true(true, 'assertion 1')
  assert.true(false, 'assertion 2')
  assert.true(false, 'assertion 3')
  assert.true(true, 'assertion 4')
})
```

The output will be:

<img width="587" alt="Screenshot 2024-02-14 at 8 55 02 AM" src="https://github.com/brandonaaron/web-test-runner-qunit/assets/13718/2c6345ce-93f2-44e4-9a29-02d5f6a4e29b">

brandonaaron replied:
Hey @rnixx, I wanted to check in to see if you've tried the latest in this PR? I'd appreciate your input on how it works for you.

rnixx replied:
@brandonaaron Sorry for my late reply, I am quite busy these days. I just tried your changes and at a first glance it looks really good now! Just one minor thing. Instead of printing 'Unknown Error' if an assertion description is missing maybe 'Assertion Error' would be more appropriate. Thanks for all the improvements you made!

brandonaaron replied:
I like the idea of using 'Assertion Error' as the fallback for a missing assertion message and got that added.

rnixx commented on README.md:
```diff
@@ -28,3 +28,46 @@ export default {
+### Progress Reporting
+
+Progress isn't reported per test but instead is all at once. This currently appears to be a limitation of this type of custom runner for @web/test-runner.
+
```
This sentence is imho a little bit confusing. WTR prints progress at least per test file.

brandonaaron replied:
Good point. I'll update it to indicate it is at best per test file.

rnixx commented on README.md:
```diff
@@ -28,3 +28,46 @@ export default {
+### Assertions that occur after a test finished
+
+This does not capture assertions that happen after the full test suite has finished running (the `runEnd` QUnit event). This could happen if there is an issue with an async test that triggers an assertion after the test suite has finished running. Perhaps a workaround could be to utilize [`assert.expect`](https://api.qunitjs.com/assert/expect/).
+
```
As pointed out previously, in such a case normally the test case itself is broken. Do you really want to mention this?

brandonaaron replied:
I think since it is something QUnit has special handling for, I'd like to call out that this adapter can't capture that error.

rnixx commented on tests/test.ts:
```diff
@@ -8,6 +8,15 @@ QUnit.todo('testing todo', (assert) => {
+
 QUnit.module('Group A', hooks => {
   // It is valid to call the same hook methods more than once.
   hooks.beforeEach(assert => {
```
If i change an assertion to fail, the beforeEach and afterEach calls of the corresponding module are marked as failed as well in the summary. Is there some information contained in the data qunit reports to avoid this?

brandonaaron replied:
This is bug on my end. I'll push up a fix shortly.

brandonaaron committed:
Apply the individual assertion passed status versus the overall test

brandonaaron committed:
Provide a default error message instead of "Unknown Error"

brandonaaron committed:
Clarify progress reporting is per file

brandonaaron committed:
Map QUnit.test to TestResult and only show the first error

This has the limitation of only showing one error per a test at a time.
In the future this behavior could be changed with a setting. However,
in order to change this behavior then each assertion likely needs to be
mapped to a TestResult.

rnixx commented:
Hi @brandonaaron, is anything missing from your point of view or is this PR ready to go + release?

brandonaaron committed:
Use a default TestResultError.message of 'Assertion Error'

This applies when an assertion does not have an associated message.

brandonaaron committed:
Tidy up some typings and config

brandonaaron committed:
Include test:open npm run script

brandonaaron committed:
Update dependencies

brandonaaron committed:
Update typescript dependencies

brandonaaron committed:
Update changelog

brandonaaron committed:
Add a couple more tests

I'm mostly using these to smoke test versus actual unit testing. It'd be
nice to be able to test that we get the failures and output we're
expecting. For now I'm just toggling values to create unit test failures
for manually checking.

brandonaaron commented:
I'm happy with how this turned out and will merge and release shortly. Ended up doing some maintenance as part of this PR too.
